File: src/main/java/xyz/mlhmz/gaspricelog/persistence/entities/ForecastGroup.java
package xyz.mlhmz.gaspricelog.persistence.entities;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

@Entity
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ForecastGroup {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID uuid;

    /**
     * Display name of the forecast group
     */
    private String groupName;

    /**
     * Gas price per kwh, varies from gas provider to gas provider
     */
    private BigDecimal gasPricePerKwh;

    /**
     * There is a calculatable kwh factor per qubicmeter that varies from gas heater to gas heater.
     */
    private BigDecimal kwhFactorPerQubicmeter;

    @OneToMany(mappedBy = "forecastGroup", fetch = FetchType.EAGER)
    List<Entry> entries;

    @OneToMany(mappedBy = "forecastGroup", fetch = FetchType.EAGER, cascade = {CascadeType.REMOVE, CascadeType.PERSIST}, orphanRemoval = true)
    List<Span> spans;
}

File: ui/src/types/ForecastGroup.ts
import z from "zod";
import { entryQuerySchema } from "./Entry";
import { spanQuerySchema } from "./Span";

export const forecastGroupMutateSchema = z.object({
	groupName: z.string(),
	gasPricePerKwh: z.number(),
	kwhFactorPerQubicmeter: z.number(),
});

export type ForecastGroupMutation = z.infer<typeof forecastGroupMutateSchema>;

export const forecastGroupQuerySchema = z.object({
	uuid: z.uuid(),
	groupName: z.string(),
	gasPricePerKwh: z.number(),
	kwhFactorPerQubicmeter: z.number(),
	entries: z.array(entryQuerySchema),
	spans: z.array(spanQuerySchema),
});

export type ForecastGroupQuery = z.infer<typeof forecastGroupQuerySchema>;
